Modulo 2 - 8. Dense + Dropout + Batch Normalization

Nallely Lizbeth Serna Rivera    A00833111

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Cargar los datos
df = pd.read_csv('student_performance_data _.csv')

# Exploración de los datos
print(df.head())
print(df.info())
print(df.describe())

# Separar características y la variable objetivo (GPA)
X = df.drop(columns=['GPA', 'StudentID'])  # Eliminar 'StudentID' ya que no aporta valor predictivo
y = df['GPA']

# Escalar los datos
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Dividir en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Función para entrenar y evaluar el modelo
def train_and_evaluate(model):
    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse', metrics=['mae'])
    
    # Entrenar el modelo
    history = model.fit(X_train, y_train, epochs=50, validation_data=(X_test, y_test), batch_size=32)
    
    # Evaluar el modelo
    test_loss, test_mae = model.evaluate(X_test, y_test)
    return test_mae, history

   StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0       1001   17       1          0                  2        19.833723   
1       1002   18       0          0                  1        15.408756   
2       1003   15       0          2                  3         4.210570   
3       1004   17       1          0                  3        10.028829   
4       1005   17       1          0                  2         4.672495   

   Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0         7         1                2                0       0      1   
1         0         0                1                0       0      0   
2        26         0                2                0       0      0   
3        14         0                3                1       0      0   
4        17         1                3                0       0      0   

   Volunteering       GPA  GradeClass  
0             0  2.929196         2.0  
1             0  3

In [13]:
# Experimento 1: Modelo con una capa oculta
model_1 = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(1)
])

mae_1, history_1 = train_and_evaluate(model_1)

Epoch 1/50
60/60 [==============================] - 2s 3ms/step - loss: 1.8729 - mae: 1.0983 - val_loss: 0.6240 - val_mae: 0.6586
Epoch 2/50
60/60 [==============================] - 0s 2ms/step - loss: 0.3090 - mae: 0.4470 - val_loss: 0.1838 - val_mae: 0.3481
Epoch 3/50
60/60 [==============================] - 0s 2ms/step - loss: 0.1589 - mae: 0.3170 - val_loss: 0.1352 - val_mae: 0.2961
Epoch 4/50
60/60 [==============================] - 0s 2ms/step - loss: 0.1249 - mae: 0.2808 - val_loss: 0.1133 - val_mae: 0.2709
Epoch 5/50
60/60 [==============================] - 0s 2ms/step - loss: 0.1033 - mae: 0.2567 - val_loss: 0.0969 - val_mae: 0.2518
Epoch 6/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0877 - mae: 0.2376 - val_loss: 0.0854 - val_mae: 0.2372
Epoch 7/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0768 - mae: 0.2229 - val_loss: 0.0772 - val_mae: 0.2250
Epoch 8/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0691 - mae: 0.21

In [5]:
# Experimento 2: Modelo con tres capas ocultas
model_2 = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(16, activation='relu'),
    tf.keras.layers.Dense(1)
])

mae_2, history_2 = train_and_evaluate(model_2)

Epoch 1/50
60/60 [==============================] - 1s 4ms/step - loss: 1.9622 - mae: 1.1421 - val_loss: 0.4160 - val_mae: 0.5161
Epoch 2/50
60/60 [==============================] - 0s 2ms/step - loss: 0.2249 - mae: 0.3742 - val_loss: 0.1613 - val_mae: 0.3177
Epoch 3/50
60/60 [==============================] - 0s 2ms/step - loss: 0.1194 - mae: 0.2748 - val_loss: 0.1064 - val_mae: 0.2568
Epoch 4/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0825 - mae: 0.2303 - val_loss: 0.0827 - val_mae: 0.2258
Epoch 5/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0658 - mae: 0.2066 - val_loss: 0.0758 - val_mae: 0.2152
Epoch 6/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0566 - mae: 0.1912 - val_loss: 0.0695 - val_mae: 0.2063
Epoch 7/50
60/60 [==============================] - 0s 2ms/step - loss: 0.0511 - mae: 0.1818 - val_loss: 0.0643 - val_mae: 0.1980
Epoch 8/50
60/60 [==============================] - 0s 1ms/step - loss: 0.0465 - mae: 0.17

In [8]:
# Experimento 3: Añadir Dropout después de cada capa oculta
model_3 = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.Dense(16, activation='relu'),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.Dense(1)
])

mae_3, history_3 = train_and_evaluate(model_3)

Epoch 1/50
60/60 [==============================] - 5s 63ms/step - loss: 2.2558 - mae: 1.2212 - val_loss: 0.5236 - val_mae: 0.6127
Epoch 2/50
60/60 [==============================] - 1s 9ms/step - loss: 0.9895 - mae: 0.7821 - val_loss: 0.4112 - val_mae: 0.5569
Epoch 3/50
60/60 [==============================] - 0s 4ms/step - loss: 0.7894 - mae: 0.7003 - val_loss: 0.2970 - val_mae: 0.4725
Epoch 4/50
60/60 [==============================] - 0s 3ms/step - loss: 0.6529 - mae: 0.6396 - val_loss: 0.2356 - val_mae: 0.4229
Epoch 5/50
60/60 [==============================] - 0s 3ms/step - loss: 0.6029 - mae: 0.6080 - val_loss: 0.2508 - val_mae: 0.4381
Epoch 6/50
60/60 [==============================] - 0s 3ms/step - loss: 0.5402 - mae: 0.5754 - val_loss: 0.1754 - val_mae: 0.3627
Epoch 7/50
60/60 [==============================] - 0s 2ms/step - loss: 0.5425 - mae: 0.5718 - val_loss: 0.1429 - val_mae: 0.3231
Epoch 8/50
60/60 [==============================] - 0s 2ms/step - loss: 0.4603 - mae: 0.5

In [6]:
# Experimento 4: Añadir Batch Normalization después de Dropout
model_4 = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(16, activation='relu'),
    tf.keras.layers.Dropout(0.3),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(1)
])

mae_4, history_4 = train_and_evaluate(model_4)

Epoch 1/50
60/60 [==============================] - 6s 4ms/step - loss: 4.8493 - mae: 1.8715 - val_loss: 3.5863 - val_mae: 1.7081
Epoch 2/50
60/60 [==============================] - 0s 3ms/step - loss: 3.4317 - mae: 1.5812 - val_loss: 2.6257 - val_mae: 1.4651
Epoch 3/50
60/60 [==============================] - 0s 2ms/step - loss: 2.4876 - mae: 1.3245 - val_loss: 1.7388 - val_mae: 1.1935
Epoch 4/50
60/60 [==============================] - 0s 3ms/step - loss: 1.8143 - mae: 1.1126 - val_loss: 1.0761 - val_mae: 0.9277
Epoch 5/50
60/60 [==============================] - 0s 2ms/step - loss: 1.3192 - mae: 0.9315 - val_loss: 0.5983 - val_mae: 0.6818
Epoch 6/50
60/60 [==============================] - 0s 2ms/step - loss: 0.9857 - mae: 0.8012 - val_loss: 0.3470 - val_mae: 0.5085
Epoch 7/50
60/60 [==============================] - 0s 2ms/step - loss: 0.8318 - mae: 0.7286 - val_loss: 0.2592 - val_mae: 0.4385
Epoch 8/50
60/60 [==============================] - 0s 3ms/step - loss: 0.7038 - mae: 0.66

In [9]:
# Comparativa de los resultados
results = pd.DataFrame({
    'Model': ['Single Dense Layer', 'Three Dense Layers', 'Dropout', 'Dropout + Batch Norm'],
    'Test MAE': [mae_1, mae_2, mae_3, mae_4]
})

print(results)

                  Model  Test MAE
0    Single Dense Layer  0.162201
1    Three Dense Layers  0.179538
2               Dropout  0.218859
3  Dropout + Batch Norm  0.154970


El modelo con Dropout + Batch Normalization tiene el menor Test MAE de 0.154970, lo que significa que este es el modelo más preciso de los cuatro en predecir el GPA de los estudiantes.